### ISSUE TITLE
Performance issue for UsdImagingDelegate::SetTime / UsdImagingDelegate::ApplyPendingUpdates

### ISSUE THREAD

**Issue by csyshing** [PixarAnimationStudios/USD#1813]
### Description of Issue

Hi,

I was working on a performance issue for production and found that `UsdImagingDelegate::SetTime()` takes a bit longer than expected to run for **the first time** when user interact the prim, for example, the first time moving a prim in Maya viewport.

The USD trace shows that [`UsdImagingDelegate::_GatherDependencies()`](https://github.com/PixarAnimationStudios/USD/blob/857ffda41f4f1553fe1019ac7c7b4f08c233a7bb/pxr/usdImaging/usdImaging/delegate.cpp#L909
) takes about 1.x~10ms to finish, in our production shot, this method is being called around 12k times, which resulted in 1.2 minutes in total (first image below: USD-20.x) / 39 second (second image below: USD-22.03):

<img width="1279" alt="USD-20 x_usdImaging_delegate_gather_dependencies_current" src="https://user-images.githubusercontent.com/1593576/159865636-aedb7a55-ee0c-4272-97a7-a7e7dce110fe.png">

<img width="1277" alt="USD-22 03_usdImaging_delegate_gather_dependencies" src="https://user-images.githubusercontent.com/1593576/159865660-3deb63cb-eb00-4865-a81d-37c87b41d553.png">

We have tried to make improvement to speed up `UsdImagingDelegate::SetTime()` call:
 - Replace the result container type to be `SdfPathSet` in [`_GatherDependencies()`](https://github.com/PixarAnimationStudios/USD/blob/release/pxr/usdImaging/usdImaging/delegate.cpp#L943-L951)
    - The existing code logic copies paths, then sorts all items, then removes duplicates, looks like using a `SdfPathSet` would be more efficient:
```
    // Propose changes below:
    SdfPathSet affectedPaths;
    for (_DependencyMap::const_iterator it = start; it != end; ++it) {
        affectedPaths.emplace(it->second);
    }
    affectedCachePaths.reserve(affectedCachePaths.size() + affectedPaths.size());
    affectedCachePaths.insert(affectedCachePaths.end(),
                              affectedPaths.begin(), affectedPaths.end());
```

 - Pre-cache path dependencies in parallel and calling `_GatherDependencies()` later would be much faster: this brings in a significant speed up and the total time drops to 10.x second:

<img width="1279" alt="usdImaging_delegate_gather_dependencies_in_parallel" src="https://user-images.githubusercontent.com/1593576/159869497-447d6849-9500-47f7-81d5-757032aa5c00.png">

As a comparison, here is the breakdown matrix based on one of our production shots with 12k prim paths to run:

| | _GatherDependencies (called from _ResyncUsdPrim and _RefreshUsdObject) | _ProcessRemoval | _Populate | _ExecuteWorkForVariabilityUpdate | SetTime (Total) |
| --- | --- | --- | --- | --- | --- |
| USD 20.x |  ~62 sec |  ~3.x sec | ~5.5 sec | ~0.5 sec | 1.2 min |
| USD 22.03 | ~30 sec | ~3.x sec | ~5.5 sec | ~0.5 sec | 39.x sec |
| Run in Parallel | ~1.x sec | ~3.x sec  | ~5.5 sec | ~0.5 sec | 10.x sec |

We are happy to submit the changes for review, but I am creating this issue to share the idea of the improvement, in case there is anything I missed.

Thanks,
Zhicheng

### Steps to Reproduce
1. Load any USD scene
2. Enable the tracing and observe the log

### System Information (OS, Hardware)

OS: CentOS 7.8

### Package Versions

Tested in USD-20.11, USD-22.03.

**Comment by jilliene** [comment]
Filed as internal issue #USD-7290

**Comment by tcauchois** [comment]
Hey Zhicheng, do you know if the SdfPathSet change by itself sped things up?  We've found that sets aren't always faster than sorted arrays, but if it's faster in this case it seems like a good change.

If your CLA is all set it would be great to see a PR for this work.

Thanks!
Tom

**Comment by csyshing** [comment]
Hi @tcauchois , I have submitted the PR here: https://github.com/PixarAnimationStudios/USD/pull/1815

I have a few tests against our production shot, the `SdfPathSet` change has a very tiny difference comparing with using `SdfPathVector`, I don't think our production shot is a good example for benchmarking this change, most of the `usd path <--> cache path` are 1 to 1 mapping so it's not as obvious as running in parallel v.s. running sequentially. I did not include the `SdfPathSet` change in my PR, I guess we will need to revisit that at another time. Thanks!
